### PR TITLE
JIT compiler for ARMv8

### PR DESCRIPTION
**Pull request by SChernykh**

### CODE REVIEW

**Comment by tevador** [comment]
Solid work.

Verification on raspberry pi is ~11x faster with JIT:
```
> ./randomx-benchmark --verify --jit --softAes
 - light memory mode (256 MiB)
 - JIT compiled mode
 - software AES mode
 - small pages mode
Initializing ...
Memory initialized in 4.41024 s
Initializing 1 virtual machine(s) ...
Running benchmark (1000 nonces) ...
Calculated result: 10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Reference result:  10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Performance: 173.841 ms per hash
```

FYI: git is complaining about trailing whitespace in a few places:
```
<stdin>:149: trailing whitespace.
        ((uint8_t*)randomx_calc_dataset_item_aarch64_prefetch - (uint8_t*)randomx_calc_dataset_item_aarch64) +
<stdin>:158: trailing whitespace.
        ) +
<stdin>:950: trailing whitespace.
        emitMemLoadFP<tmp_reg_fp>(src, instr, code, k);
<stdin>:973: trailing whitespace.
        emitMemLoadFP<tmp_reg_fp>(src, instr, code, k);
<stdin>:1003: trailing whitespace.
        emitMemLoadFP<tmp_reg_fp>(src, instr, code, k);
```

File: src/jit_compiler_a64.cpp [hunk 1]
```diff
@@ -0,0 +1,1068 @@
...
+}
+
+namespace randomx {
+
+static const size_t CodeSize = ((uint8_t*)randomx_init_dataset_aarch64_end) - ((uint8_t*)randomx_program_aarch64);
+static const size_t MainLoopBegin = ((uint8_t*)randomx_program_aarch64_main_loop) - ((uint8_t*)randomx_program_aarch64);
+static const size_t PrologueSize = ((uint8_t*)randomx_program_aarch64_vm_instructions) - ((uint8_t*)randomx_program_aarch64);
+static const size_t ImulRcpLiteralsEnd = ((uint8_t*)randomx_program_aarch64_imul_rcp_literals_end) - ((uint8_t*)randomx_program_aarch64);
+
+static const size_t CalcDatasetItemSize =
```
Comment (tevador) [inline src/jit_compiler_a64.cpp]: This should depend on `RANDOMX_PROGRAM_SIZE`.
I get segfaults when I increase the program size.

**Comment by rbthomp** [comment]
Here is what I'm getting on my ARMV8 CPU. 

```
./randomx-benchmark --verify --jit
RandomX benchmark v1.1.0
 - light memory mode (256 MiB)
 - JIT compiled mode
 - hardware AES mode
 - small pages mode
Initializing ...
Initializing 1 virtual machine(s) ...
Running benchmark (1000 nonces, iteration 0) ...
Calculated result: 10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Reference result:  10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Performance: 59.3619 ms per hash
```
```
/randomx-benchmark --verify --nonces 1000 --jit --threads 8
RandomX benchmark v1.1.0
 - light memory mode (256 MiB)
 - JIT compiled mode
 - hardware AES mode
 - small pages mode
Initializing ...
Initializing 8 virtual machine(s) ...
Running benchmark (1000 nonces, iteration 0) ...
Calculated result: 10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Reference result:  10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Performance: 12.5699 ms per hash
```
```
./randomx-benchmark --mine --threads 8 --init 8 --jit
RandomX benchmark v1.1.0
 - full memory mode (2080 MiB)
 - JIT compiled mode
 - hardware AES mode
 - small pages mode
Initializing (8 threads) ...
Dataset loaded in 7.47393 s
Initializing 8 virtual machine(s) ...
Running benchmark (1000 nonces, iteration 0) ...
Calculated result: 10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Reference result:  10b649a3f15c7c7f88277812f2e74b337a0f20ce909af09199cccb960771cfa1
Performance: 240.109 hashes per second
```

I generally get better performance when using GCC-8 on ARMV8, but I get a segmentfault after initialization on GCC-8.

File: src/jit_compiler_a64.cpp (hunk 1 above)
Comment (SChernykh) [reply src/jit_compiler_a64.cpp]: Fixed in https://github.com/tevador/RandomX/pull/125/commits/ef19e298e123330ec3234a5ba1054bff82e71d7d

**Comment by SChernykh** [comment]
@rbthomp I've tested it with gcc 7.4.0, gcc 8.3.0 on an ARM server and clang 8.0.1 on an Android phone, and I didn't get any crashes.

**Pull request merged**